habitat_baselines/train_ppo.py

```diff
@@ -224,137 +225,184 @@ def run_training():
     count_steps = 0
     count_checkpoints = 0
 
-    for update in range(args.num_updates):
-        if args.use_linear_lr_decay:
-            update_linear_schedule(
-                agent.optimizer, update, args.num_updates, args.lr
-            )
+    tensorboard_dir = args.tensorboard_dir
 
-        agent.clip_param = args.clip_param * (1 - update / args.num_updates)
+    writer_kwargs = dict(
+        log_dir=tensorboard_dir, purge_step=count_steps, flush_secs=30
+    )
+    with (tensorboard.SummaryWriter(**writer_kwargs)) as writer:
+        for update in range(args.num_updates):
+            if args.use_linear_lr_decay:
+                update_linear_schedule(
+                    agent.optimizer, update, args.num_updates, args.lr
+                )
 
-        for step in range(args.num_steps):
-            t_sample_action = time()
-            # sample actions
-            with torch.no_grad():
-                step_observation = {
-                    k: v[step] for k, v in rollouts.observations.items()
-                }
+            agent.clip_param = args.clip_param * (
+                1 - update / args.num_updates
+            )
 
-                (
-                    values,
-                    actions,
-                    actions_log_probs,
-                    recurrent_hidden_states,
-                ) = actor_critic.act(
-                    step_observation,
-                    rollouts.recurrent_hidden_states[step],
-                    rollouts.masks[step],
-                )
-            pth_time += time() - t_sample_action
+            for step in range(args.num_steps):
+                t_sample_action = time()
+                # sample actions
+                with torch.no_grad():
+                    step_observation = {
+                        k: v[step] for k, v in rollouts.observations.items()
+                    }
+
+                    (
+                        values,
+                        actions,
+                        actions_log_probs,
+                        recurrent_hidden_states,
+                    ) = actor_critic.act(
```

Comment:
Let's fix readability here.

Reply:
I think the `train_ppo` code in general needs a revamp and better abstractions, I am at fault here as this was written in a hasty manner. We should discuss how to restructure this.